src/main/java/io/jenkins/jenkins/plugins/okhttp/api/JenkinsOkHttpClient.java

```diff
@@ -0,0 +1,172 @@
+        OkHttpClient.Builder reBuild = httpClient.newBuilder();
+        if (Jenkins.get().proxy != null) {
+            final ProxyConfiguration proxy = Jenkins.get().proxy;
+            Proxy proxyServer = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxy.name, proxy.port));
+            if (proxy.getUserName() != null) {
+                Authenticator proxyAuthenticator = new Authenticator() {
+                    @Override public Request authenticate(Route route, Response response) throws IOException {
+                        String credential = Credentials.basic(proxy.getUserName(), proxy.getPassword());
+                        return response.request().newBuilder()
+                                       .header("Proxy-Authorization", credential)
+                                       .build();
+                    }
+                };
+                reBuild.proxyAuthenticator(proxyAuthenticator);
+            }
+
+            // TODO: https://github.com/square/okhttp/issues/3787
+            // not sure if this is correct, needs manual testing
+            if (proxy.name.startsWith("https")) {
+                reBuild.socketFactory(SSLSocketFactory.getDefault());
+            }
+            ProxySelector proxySelector = new ProxySelector() {
+                @Override
+                public List<Proxy> select(URI uri) {
+                    final List<Proxy> proxies = new ArrayList<>(1);
+                    String host = uri.getHost();
+                    for (Pattern p : proxy.getNoProxyHostPatterns()) {
+                        if (p.matcher(host).matches()) {
+                            proxies.add(Proxy.NO_PROXY);
+                            return proxies;
+                        }
+                    }
+                    proxies.add(proxyServer);
+                    return proxies;
+                }
+
+                @Override
+                public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+                    LOGGER.log(Level.WARNING, "Proxy connection failed", ioe);
+                }
+            };
+            reBuild.proxySelector(proxySelector);
+        } else {
+            reBuild.proxy(Proxy.NO_PROXY);
+        }
+
+        if (acceptAnyCertificate) {
+            try {
+                applyInsecureConfiguration(reBuild);
+            } catch (NoSuchAlgorithmException | KeyManagementException e) {
+                LOGGER.log(Level.WARNING, "Cannot apply insecure trust all certs configuration to okhttp", e);
+            }
+        }
+        return reBuild;
+    }
+
+    /**
+     * Updates non-proxy-hosts on the given client (returns a new builder object as a rebuild of the given one).
+     * @param httpClient the base client
+     * @param nonProxyHosts list of non proxy hosts patterns
+     * @return the builder for the new client
+     */
+    public static OkHttpClient.Builder withNonProxyHosts(OkHttpClient httpClient, List<String> nonProxyHosts) {
```

Comment:
I'm wondering why we would need different nonProxyHosts from the list declared on Jenkins instance